### PR TITLE
feat(collections): auto-generate collection.md with maturity filtering

### LINKED ISSUE #1256: collection.md: auto-generation, maturity safety, and custom header support
## Context

This issue was raised during review of PR #1207, triggered by a question about why `collections/hve-core-all.collection.md` was missing the newly added skill, and a secondary observation that the `.collection.md` files are not auto-generated despite the plugin README files being auto-generated.

Cited discussion comments: https://github.com/microsoft/hve-core/pull/1207#discussion_r3018383074

---

## Current Behavior (by design, but worth revisiting)

The three files involved in producing each published surface (plugin README and extension marketplace README) are:

| File | Authored by | Used for |
|---|---|---|
| `collections/{id}.collection.yml` | Contributor (manual) | Structured metadata, item list with maturity, `notice:` field |
| `collections/{id}.collection.md` | Contributor (manual) | Prose body: overview paragraph, feature/agent/skill bullet lists |
| `plugins/{id}/README.md` | Auto-generated by `npm run plugin:generate` | Copilot CLI marketplace page |
| `extension/README.{id}.md` | Auto-generated by `Prepare-Extension.ps1` | VS Code Marketplace page |

**Generation flow:**

- `plugins/{id}/README.md` is fully auto-generated: it takes the `yml` `name`/`description`/`notice` fields for the header, inserts the entire `collection.md` content as an `## Overview` section, then appends auto-built artifact tables (Agents, Commands, Instructions, Skills) from the `yml` items list.
- `extension/README.{id}.md` follows the same pattern using `README.template.md` tokens: `{{BODY}}` = full `collection.md`, `{{ARTIFACTS}}` = auto-built artifact tables from `yml` items.
- The `hve-core-all.collection.yml` **is** auto-populated by `Update-HveCoreAllCollection` (scanning all other collections), but `hve-core-all.collection.md` is **not** auto-updated.

**Root cause of the PR confusion:** When a contributor adds a new artifact to `collection.yml`, the artifact tables in both generated READMEs update automatically. However, the named bullet lists inside `collection.md` (e.g., `- **OWASP MCP Top 10** — ...`) are hand-authored and silently fall out of sync. There is no linting or validation step that detects this gap.

---

### 1. collection.md prose body is not auto-generated

The bullet list sections in `collection.md` (listing agents, subagents, skills with short descriptions) largely duplicate data already present as YAML frontmatter `description:` fields on each artifact. When a contributor adds a new artifact to `collection.yml` and does not update `collection.md`, the prose body becomes stale.

The ambiguity is worsened because `Generate-Plugins.ps1` **does** auto-update `hve-core-all.collection.yml` with newly discovered artifacts, creating a justified expectation that the companion `.md` would also update. It does not.

### 2. Extension README artifact tables do not filter by maturity channel

`New-CollectionReadme` in `Prepare-Extension.ps1` iterates all items from the collection manifest without applying the `Channel` maturity filter. The function call site (around line 320) also does not pass the active channel. This means the **stable extension's** marketplace README (`extension/README.{id}.md`) lists artifact names from `experimental` and `preview` items alongside stable ones.

The `package.json` artifact registration **does** filter by maturity (via `Get-CollectionArtifacts` with `AllowedMaturities`), so the agent picker in stable VS Code correctly omits experimental agents — but the README visible to users on the marketplace says otherwise.

### 3. collection.md body prose may name experimental artifacts

Because `collection.md` is hand-authored prose that is inserted verbatim into both the plugin README and extension README, any artifact names mentioned in the prose (e.g., `- **Security Planner** — ...`) appear in the stable extension's marketplace page. There is no mechanism to conditionally exclude those names from the stable surface.

### 4. No custom header/intro slot separate from the notice

The `yml` `notice:` field serves callout blocks (CAUTION, WARNING, etc.), but there is no supported way to add per-collection introductory text (e.g., a one-paragraph why/when of the collection) that appears before the prose body and is distinct from the warning-style `notice`.

---

### Option A — Auto-generate collection.md entirely from yml + artifact frontmatter

Generate the entire `collection.md` prose from data already in `collection.yml` (item paths, kinds) plus `description:` fields read from each artifact's YAML frontmatter. The script writes `collection.md` as a generated file.

- **Pro:** Single source of truth; always in sync; removes contributor burden.
- **Con:** Loses ability to write custom non-tabular prose (e.g., narrative context). All prose must come from artifact `description:` fields, which are intentionally terse one-liners. The generated bullet sections would closely mirror the artifact tables already present below them, adding little value.
- **Con:** `hve-core-all.collection.md` has deliberate curated prose introducing the mega-collection that cannot be derived from yml alone.
- **Verdict:** Suitable only if the body is intentionally reduced to structured data; doesn't map well to the current narrative style.

### Option B — Add `header:` YAML field to collection.yml; keep collection.md for body prose

Add a multiline `header:` field to `collection.yml` (analogous to `notice:`). Scripts inject it before the `collection.md` body. The body prose sections (feature lists, agent descriptions) remain hand-authored in `collection.md`, but are validated.

- **Pro:** Explicit, structured header with version control; easy to see in yml review.
- **Con:** Introduces a second prose field in yml alongside the existing `notice:`. For collections that don't need a header, the field is just omitted. The core sync problem (body prose vs. yml items) is not addressed.

### Option C — Template sections within collection.md (recommended for discussion)

Give `collection.md` an optional YAML frontmatter block that carries a `header:` string. Below the frontmatter, the body is partitioned into:
1. A hand-authored `<!-- hve:intro -->` block (intro paragraph, preserved as-is).
2. Auto-generated `<!-- hve:agents -->`, `<!-- hve:skills -->` etc. blocks that `Generate-Plugins.ps1` regenerates from yml items + each artifact's `description:` frontmatter.

The script would regenerate only the tagged blocks, leaving the intro unchanged. The plugin and extension READMEs pull the full `collection.md` as before.

- **Pro:** Keeps customizable intro prose; auto-syncs the artifact bullet lists; container tags make it machine-readable.
- **Con:** More complex script logic. Contributors must not edit inside auto-generated blocks. Tag convention needs documentation.

### Option D — Move collection.md body to yml (`body:` multiline field); auto-generate the artifact lists appended into it

All prose lives in `collection.yml` as `body:` multiline text. Artifact bullet sections are appended automatically by the script at generation time. `collection.md` becomes a generated file and is removed from contributor ownership.

- **Pro:** One file to edit (yml). The generator always produces a complete, correct `collection.md`.
- **Con:** Multiline YAML prose is harder to write/review than a markdown file. Loss of markdown preview ergonomics.

---

## Immediate Bug to Fix (independent of design choice)

Regardless of which option is chosen for the auto-generation design, `New-CollectionReadme` in `Prepare-Extension.ps1` should receive and apply the active `Channel` maturity filter when building artifact tables. The `$Channel` parameter is already available in the calling context and `Get-CollectionArtifacts` already implements the correct filtering logic. Passing the filtered collection (or filtered artifact lists) into `New-CollectionReadme` would close the stable-extension-lists-experimental-agents gap.

---

## Questions for Team Discussion

1. Is auto-generation of `collection.md` body desirable, or is the hand-authored narrative a deliberate quality bar we want to maintain?
2. If auto-generation is adopted (Option C or D), should `-Refresh` on `Generate-Plugins.ps1` also regenerate `collection.md` blocks, or should there be a separate `npm run collections:sync` command?
3. For the `header:` slot: should it live in `yml` (Option B) or in `collection.md` frontmatter (Option C)? The `yml` location is simpler and follows the precedent of `notice:`.
4. Should the stable extension README for security/experimental collections explicitly state which items require the Pre-Release channel, rather than just omitting them from the listing?

---

## Related

- PR #1207 — context for this issue
- `scripts/plugins/Modules/PluginHelpers.psm1` — `New-PluginReadmeContent`, plugin README generation
- `scripts/extension/Prepare-Extension.ps1` — `New-CollectionReadme`, extension README generation + `Get-CollectionArtifacts` (maturity filtering already implemented for package.json but not README)

### PR DESCRIPTION
**Pull request by WilliamBerryiii**
## Description

Added auto-generation of `collection.md` artifact tables from YAML manifests with channel-based maturity filtering and sentinel marker support. The generation engine in *Prepare-Extension.ps1* gained a `-Channel` parameter that controls which maturity levels appear in output, and `New-CollectionReadme` was rewritten to parse marker-delimited sections, filter items by maturity, and write back updated content preserving hand-written introductions and footers. Collection validation now detects mismatched or reversed markers as warnings for backward compatibility. The encoding for `Set-ContentIfChanged` was corrected to UTF-8 without BOM. Nineteen new Pester tests cover maturity filtering, template marker handling, and `Split-CollectionMdByMarkers` edge cases.

## Related Issue(s)

Closes #1256

## Type of Change

Select all that apply:

**Code & Documentation:**

* [ ] Bug fix (non-breaking change fixing an issue)
* [x] New feature (non-breaking change adding functionality)
* [ ] Breaking change (fix or feature causing existing functionality to change)
* [ ] Documentation update

**Infrastructure & Configuration:**

* [ ] GitHub Actions workflow
* [ ] Linting configuration (markdown, PowerShell, etc.)
* [ ] Security configuration
* [ ] DevContainer configuration
* [ ] Dependency update

**AI Artifacts:**

* [ ] Reviewed contribution with `prompt-builder` agent and addressed all feedback
* [ ] Copilot instructions (`.github/instructions/*.instructions.md`)
* [ ] Copilot prompt (`.github/prompts/*.prompt.md`)
* [ ] Copilot agent (`.github/agents/*.agent.md`)
* [ ] Copilot skill (`.github/skills/*/SKILL.md`)

**Other:**

* [x] Script/automation (`.ps1`, `.sh`, `.py`)
* [ ] Other (please describe):

## Testing

**Automated testing:**

- **Pester test suite:** 2002 tests executed — 1935 passed, 0 failed, 2 skipped (7m40s)
- **New test coverage:** 19 new tests in *Prepare-Extension.Tests.ps1* (+233 lines) and *Validate-Collections.Tests.ps1* (+157 lines)
  - Maturity filtering: excludes experimental when stable-only, includes experimental when allowed, excludes deprecated regardless
  - Template markers: preserves intro and replaces marker section, writes back updated artifacts, handles files without markers, preserves footer content
  - `Split-CollectionMdByMarkers`: no markers, empty string throws, correct intro/footer parsing, partial markers, reversed markers, duplicate markers
- **Lint validation:** 100% compliance across markdown, PowerShell, YAML, frontmatter, links, spelling, and skill structure checks
- **Plugin generation:** All 14 plugins regenerated successfully — 116 agents, 143 commands, 245 instructions, 30 skills, 0 lint errors

**Manual testing:** Not performed.

### Required Checks

* [x] Documentation is updated (if applicable)
* [x] Files follow existing naming conventions
* [x] Changes are backwards compatible (if applicable)
* [x] Tests added for new functionality (if applicable)

### Required Automated Checks

The following validation commands must pass before merging:

* [x] Markdown linting: `npm run lint:md`
* [x] Spell checking: `npm run spell-check`
* [x] Frontmatter validation: `npm run lint:frontmatter`
* [x] Skill structure validation: `npm run validate:skills`
* [x] Link validation: `npm run lint:md-links`
* [x] PowerShell analysis: `npm run lint:ps`

### CODE REVIEW

File: scripts/extension/Prepare-Extension.ps1 [hunk 1]
```diff
@@ -478,6 +546,17 @@ function New-CollectionReadme {
         $null = $artifactSections.AppendLine()
     }
 
+    # Write back updated artifact section into collection.md when markers are present
```
Comment (katriendg) [inline scripts/extension/Prepare-Extension.ps1:466]: ### 🔴 Comment 1 — Critical Gap: `collection.md` writeback is only wired to `extension:prepare`, not `plugin:generate`

> The sentinel-marker writeback that updates the artifact table in `collection.md` runs inside `New-CollectionReadme`, which is called only from `Invoke-ExtensionCollectionsGeneration` — and that function is only invoked by `npm run extension:prepare` (and `:prerelease`).
>
> `npm run plugin:generate` calls `Generate-Plugins.ps1`, which **never calls `Prepare-Extension.ps1`** and never writes back to `collection.md`.
>
> This means:
>
> * A contributor who adds an item to `ado.collection.yml` and runs `npm run plugin:generate` (as the docs instruct) will see the `plugins/ado/README.md` update correctly, but `collections/ado.collection.md` will remain stale — the artifact table between the markers will not reflect the new item.
> * The `collection.md` only refreshes when someone runs `extension:prepare`, which contributors won't do unless they're on the extension release path.
> * `.github/copilot-instructions.md` line 107 currently states: _"run `npm run plugin:generate` to regenerate plugin outputs under `plugins/` and update the artifact sections in `collection.md` files."_ This is incorrect — `plugin:generate` does not update `collection.md`.
>
> **Recommended fix**: Move the `collection.md` writeback into `Generate-Plugins.ps1` (or a shared helper), so that `plugin:generate` is the single command that keeps both `plugins/*/README.md` and `collections/*.collection.md` in sync. Alternatively, add a dedicated `npm run collection:sync` script and update all contributor-facing docs to reference it as a required step after editing a collection YAML.

File: scripts/extension/Prepare-Extension.ps1 [hunk 2]
```diff
@@ -64,6 +64,10 @@ $ErrorActionPreference = 'Stop'
 Import-Module (Join-Path $PSScriptRoot "../lib/Modules/CIHelpers.psm1") -Force
 Import-Module (Join-Path $PSScriptRoot "../collections/Modules/CollectionHelpers.psm1") -Force
 
+# Auto-generation marker constants shared across Split-CollectionMdByMarkers and New-CollectionReadme
+$script:CollectionMdBeginMarker = '<!-- BEGIN AUTO-GENERATED ARTIFACTS -->'
```
Comment (katriendg) [inline scripts/extension/Prepare-Extension.ps1]: ### 🟡 Comment 3 — Suggestion: marker string literals duplicated across two files

* File: `scripts/extension/Prepare-Extension.ps1` (lines 68-69) and `scripts/plugins/Modules/PluginHelpers.psm1`
* Category: Maintainability
* Severity: Suggestion

> Once Comment 1 is fixed, the marker strings `<!-- BEGIN AUTO-GENERATED ARTIFACTS -->` and `<!-- END AUTO-GENERATED ARTIFACTS -->` will be defined in two separate files. If the strings ever change, both must be updated in sync — a silent breakage risk.
>
> Both files already import `CollectionHelpers.psm1`. Exporting the constants from that module (or a new `MarkerConstants.psm1`) would give a single source of truth. At minimum, add a comment in each file cross-referencing the other.

**Review by katriendg — COMMENTED** [review]
Left a few comments, I believe ensuring the plugin:generate takes care of generating the `<collectionid>.collection.md` file is a good approach, and we could have the Extension prepare regenerate what is needed based on maturity filtering.

Plugins always take everything without maturity filtering, so that in itself so the more I reflect on this, the more I think we need a generation for the plugins (the default), and regenerate one for stable/pre-release when packaging the extension.

Also this will impact the following documentation addition:

### 🟡 Comment 2 — Documentation gap: contributor guide has no mention of the two-zone `collection.md` convention

* File: `docs/contributing/ai-artifacts-common.md` and `.github/copilot-instructions.md`
* Category: Documentation
* Severity: **Should fix before merge**

> `docs/contributing/ai-artifacts-common.md` (the primary contributor guide for AI artifacts) extensivelydocuments `*.collection.yml` structure, maturity values, and plugin generation — but has no mention of:
>
> 1. The `<!-- BEGIN AUTO-GENERATED ARTIFACTS -->` / `<!-- END AUTO-GENERATED ARTIFACTS -->` marker format.
> 2. That `collection.md` has two zones (hand-authored intro vs auto-generated table) and contributors must only edit the intro.
> 3. What command to run to refresh the artifact table after editing a YAML.
> 4. How to seed a new collection with the correct marker structure.
>
> `.github/copilot-instructions.md` line 104 says _"Only edit the intro zone; the artifact table is overwritten by `plugin:generate`"_ — this was added during the review session but the underlying `plugin:generate` wiring gap (Comment 2 above) means this statement is currently inaccurate.
>
> A `### Collection Markdown Auto-Generation` subsection should be added to `docs/contributing/ai-artifacts-common.md` explaining the marker format, the two-zone rule, and the correct command to sync the artifact table. This section should be updated once Comment 2 is resolved so the documented command is accurate.

File: scripts/collections/Validate-Collections.ps1 [hunk 3]
```diff
@@ -23,6 +23,10 @@ $ErrorActionPreference = 'Stop'
 Import-Module (Join-Path $PSScriptRoot 'Modules/CollectionHelpers.psm1') -Force
 Import-Module (Join-Path $PSScriptRoot '../lib/Modules/CIHelpers.psm1') -Force
 
+# Auto-generation marker constants for collection.md validation
+$script:CollectionMdBeginMarker = '<!-- BEGIN AUTO-GENERATED ARTIFACTS -->'
+$script:CollectionMdEndMarker = '<!-- END AUTO-GENERATED ARTIFACTS -->'
+
```
[the remark this answers is not in the document]
Comment (WilliamBerryiii) [reply scripts/collections/Validate-Collections.ps1]: Addressed — marker constants (`$CollectionMdBeginMarker`, `$CollectionMdEndMarker`) are now centralized in `CollectionHelpers.psm1` as module-scoped variables exported via `Export-ModuleMember -Variable`. Both `Validate-Collections.ps1` and `Prepare-Extension.ps1` consume them from the shared module, eliminating all duplication.

File: scripts/extension/Prepare-Extension.ps1 [hunk 4]
```diff
@@ -368,6 +379,40 @@ function Get-ArtifactDescription {
...
 
+function Split-CollectionMdByMarkers {
+    <#
+    .SYNOPSIS
+        Splits collection.md content at auto-generation markers.
+    #>
+    param(
+        [Parameter(Mandatory)]
+        [string]$Content
+    )
```
[the remark this answers is not in the document]
Comment (WilliamBerryiii) [reply scripts/extension/Prepare-Extension.ps1]: Addressed — `Split-CollectionMdByMarkers` now includes `[CmdletBinding()]`, `[OutputType([hashtable])]`, and complete comment-based help with `.SYNOPSIS`, `.DESCRIPTION`, `.PARAMETER`, `.OUTPUTS`, and `.EXAMPLE` blocks, following the project's PowerShell conventions.

File: scripts/extension/Prepare-Extension.ps1 (hunk 1 above)
Comment (WilliamBerryiii) [reply scripts/extension/Prepare-Extension.ps1:466]: Addressed — collection.md writeback logic has been added to `Generate-Plugins.ps1` so that running `npm run plugin:generate` now updates the relevant `collections/*.collection.md` files. This closes the critical gap where contributors running only the plugin generation pipeline would see stale collection.md content. The writeback uses the same shared marker constants from `CollectionHelpers.psm1`.

File: scripts/extension/Prepare-Extension.ps1 (hunk 2 above)
Comment (WilliamBerryiii) [reply scripts/extension/Prepare-Extension.ps1]: Addressed — this is resolved by the same change described in the response to the review comment above. The marker string literals (`<!-- BEGIN AUTO-GENERATED ARTIFACTS -->` / `<!-- END AUTO-GENERATED ARTIFACTS -->`) are now defined once in `CollectionHelpers.psm1` as exported module variables (`$CollectionMdBeginMarker`, `$CollectionMdEndMarker`), providing a single source of truth across all consuming scripts.

**Review by katriendg — APPROVED** [review]
Thanks, I think is should be good to merge. Just noticed you need to regenerate the plugins.

**Review by rezatnoMsirhC — APPROVED** [review]

**Review by bindsi — APPROVED** [review]

**Review by rezatnoMsirhC — APPROVED** [review]
